lib/ruby_lsp/requests/completion.rb

```diff
@@ -43,265 +45,65 @@ def provider
-          filter_text.prepend("::")
-        end
-
-        # If the user is searching for a constant inside the current namespace, then we prefer completing the short name
-        # of that constant. E.g.:
-        #
-        # module Foo
-        #  class Bar
-        #  end
-        #
-        #  Foo::B # --> completion inserts `Bar` instead of `Foo::Bar`
-        # end
-        @nesting.each do |namespace|
-          prefix = "#{namespace}::"
-          shortened_name = insertion_text.delete_prefix(prefix)
-
-          # If a different entry exists for the shortened name, then there's a conflict and we should not shorten it
-          conflict_name = "#{@nesting.join("::")}::#{shortened_name}"
-          break if real_name != conflict_name && @index[conflict_name]
-
-          insertion_text = shortened_name
-
-          # If the user is typing a fully qualified name `Foo::Bar::Baz`, then we should not use the short name (e.g.:
-          # `Baz`) as filtering. So we only shorten the filter text if the user is not including the namespaces in their
-          # typing
-          filter_text.delete_prefix!(prefix) unless incomplete_name.start_with?(prefix)
-        end
-
-        # When using a top level constant reference (e.g.: `::Bar`), the editor includes the `::` as part of the filter.
-        # For these top level references, we need to include the `::` as part of the filter text or else it won't match
-        # the right entries in the index
-        Interface::CompletionItem.new(
-          label: real_name,
-          filter_text: filter_text,
-          text_edit: Interface::TextEdit.new(
-            range: range_from_node(node),
-            new_text: insertion_text,
-          ),
-          kind: kind,
-          label_details: Interface::CompletionItemLabelDetails.new(
-            description: entries.map(&:file_name).join(","),
-          ),
-          documentation: markdown_from_index_entries(real_name, entries),
-        )
-      end
-
-      # Check if there are any conflicting names for `entry_name`, which would require us to use a top level reference.
-      # For example:
-      #
-      # ```ruby
-      # class Bar; end
-      #
-      # module Foo
-      #   class Bar; end
-      #
-      #   # in this case, the completion for `Bar` conflicts with `Foo::Bar`, so we can't suggest `Bar` as the
-      #   # completion, but instead need to suggest `::Bar`
-      #   B
-      # end
-      # ```
-      sig { params(entry_name: String).returns(T::Boolean) }
-      def top_level?(entry_name)
-        @nesting.length.downto(0).each do |i|
-          prefix = T.must(@nesting[0...i]).join("::")
-          full_name = prefix.empty? ? entry_name : "#{prefix}::#{entry_name}"
-          next if full_name == entry_name
-
-          return true if @index[full_name]
-        end
-
-        false
+      sig { override.returns(ResponseType) }
+      def response
+        @dispatcher.dispatch_once(@target)
```

Comment:
This line is failing because `@dispatcher` is `nil` on certain conditions due to the early returns in `initialize`.

I think we should move most of the code in the `initialize` method to `response` and ensure that the key instance variables are always set.

Reply:
I prefer not moving most of them into `response` as it means we'd need to initialise listeners there too, which doesn't align with other requests.
I've rearranged the ordering in `initialize` a bit instead so all the ivars used in `response` should always be defined.